olp-cpp-sdk-core/tests/thread/unittest/SyncQueueTest.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+#include <olp/core/thread/SyncQueue.h>
+
+using SyncTaskType = std::function<void()>;
+using SyncQueueFifo = olp::thread::SyncQueueFifo<SyncTaskType>;
+
+using SyncIncType = std::function<void(size_t)>;
+using SyncQueueIncFifo = olp::thread::SyncQueueFifo<SyncIncType>;
+
+using SharedQueueType = std::shared_ptr<std::string>;
+using SyncQueueShared = olp::thread::SyncQueueFifo<SharedQueueType>;
+
+using namespace ::testing;
+using namespace std::chrono;
+
+TEST(SyncQueueTest, initialize) {
+  SCOPED_TRACE("Default initialized not closed");
+  SyncQueueFifo sync_queue;
+  bool check = false;
+  sync_queue.Push([&check]() { check = true; });
+
+  bool closed = false;
+  SyncTaskType element;
+  sync_queue.Pull(element, closed);
+  ASSERT_TRUE(element) << "Push() or Pull() failed.";
+  element();
+
+  EXPECT_TRUE(sync_queue.Empty());
+  EXPECT_FALSE(closed);
+  EXPECT_TRUE(check);
+}
+
+TEST(SyncQueueTest, push) {
+  {
+    SCOPED_TRACE("Push supports rvalues");
+    SyncQueueShared sync_queue;
+
+    auto string = std::make_shared<std::string>("rvalue");
+    sync_queue.Push(std::move(string));
+
+    EXPECT_FALSE(sync_queue.Empty());
+    EXPECT_FALSE(string) << "string should be moved";
+  }
+  {
+    SCOPED_TRACE("Push supports rvalues");
+    SyncQueueShared sync_queue;
+
+    auto string = std::make_shared<std::string>("lvalue");
+    sync_queue.Push(string);
+
+    EXPECT_FALSE(sync_queue.Empty());
+    EXPECT_TRUE(string && !string->empty());
+  }
+}
+
+TEST(SyncQueueTest, pull) {
```

Comment:
Divide into 2 test cases - for open queue and closed queue

Reply:
See above.